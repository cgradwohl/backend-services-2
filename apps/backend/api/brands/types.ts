import { ApiRequestContext } from "~/lib/lambda-response";
import {
  IApiBrandsGetResponse,
  IApiBrandsListResponse,
  IApiBrandsPostRequest,
  IApiBrandsPostResponse,
  IApiBrandsPutRequest,
  IApiBrandsPutResponse,
} from "~/types.public";

export type ApiBrandsResponse =
  | IApiBrandsGetResponse
  | IApiBrandsListResponse
  | IApiBrandsPostResponse
  | IApiBrandsPutResponse;

export type GetFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsGetResponse;
}>;

export type ListFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsListResponse;
}>;

export type PostRequestBody = IApiBrandsPostRequest;

export type PostFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsPostResponse;
}>;

export type PutRequestBody = IApiBrandsPutRequest;

export type PutFn = (
  context: ApiRequestContext
) => Promise<{
  body: IApiBrandsPutResponse;
}>;

export type RemoveFn = (
  context: ApiRequestContext
) => Promise<{
  status: 204;
}>;
