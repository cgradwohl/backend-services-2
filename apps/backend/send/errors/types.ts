import makeError from "make-error";
import { getErrorMessage } from "./lib";

// only status codes that indicate the service did not process the request should be retried

export enum SendErrorType {
  ALREADY_EXISTS = "ALREADY_EXISTS",
  FAILED_PRECONDITION = "FAILED_PRECONDITION",
  INTERNAL = "INTERNAL",
  INVALID_ARGUMENT = "INVALID_ARGUMENT",
  NOT_FOUND = "NOT_FOUND",
  RESOURCE_EXHAUSTED = "RESOURCE_EXHAUSTED",
  UNAVAILABLE = "UNAVAILABLE",
  UNKNOWN = "UNKNOWN",
}
type RetryableSendErrorType = Extract<
  SendErrorType,
  | SendErrorType.INTERNAL
  | SendErrorType.NOT_FOUND
  | SendErrorType.RESOURCE_EXHAUSTED
  | SendErrorType.UNAVAILABLE
  | SendErrorType.UNKNOWN
>;
type NonRetryableSendErrorType = Extract<
  SendErrorType,
  | SendErrorType.ALREADY_EXISTS
  | SendErrorType.FAILED_PRECONDITION
  | SendErrorType.INVALID_ARGUMENT
>;

const BaseError = makeError("BaseError");

export abstract class SendError extends BaseError {
  abstract retryable: boolean;
  abstract type: SendErrorType;
  public detail?: Record<string, string>;

  constructor(params: {
    error: unknown;
    type: SendErrorType;
    detail?: Record<string, string>;
  }) {
    super(getErrorMessage(params.error));

    if (params?.detail) {
      this.detail = params?.detail ?? (params?.error as any).detail;
    }

    if ((params.error as SendError)?.detail) {
      this.detail = {
        ...(params?.error as SendError)?.detail,
        ...params?.detail, // the last error to throw takes precedence
      };
    }
  }
}

export class NonRetryableSendError extends SendError {
  public retryable: false;
  public type: NonRetryableSendErrorType;

  constructor(params: {
    error: unknown;
    type: NonRetryableSendErrorType;
    detail?: Record<string, string>;
  }) {
    super(params);
    this.retryable = false;
    this.type = params.type;
  }
}

export class RetryableSendError extends SendError {
  public retryable: true;
  public type: RetryableSendErrorType;

  constructor(params: {
    error: unknown;
    type: RetryableSendErrorType;
    detail?: Record<string, string>;
  }) {
    super(params);
    this.retryable = true;
    this.type = params.type;
  }
}
