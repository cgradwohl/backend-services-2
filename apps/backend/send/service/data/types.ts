export type FilePathGenerator = (params: {
  id: string;
  name: SendDataEntity;
}) => `${string}/${string}_${string}.json`;

export interface SendDyanmoItem {
  created: string;
  updated: string;
  workspaceId: string;
}

export enum SendDataEntity {
  message = "message",
  request = "request",
  sequence = "sequence",
  sequenceAction = "sequence-action",
  scheduleJob = "schedule-job",
}
