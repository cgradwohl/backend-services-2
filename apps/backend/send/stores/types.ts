export type FilePathGenerator = ({
  requestId,
  messageId,
}: {
  requestId?: string;
  messageId?: string;
}) => string;

export type S3Get<T> = (payload: { filePath: string }) => Promise<T>;
export type S3Put<T> = (payload: {
  requestId?: string;
  messageId?: string;
  json: T;
}) => Promise<{ filePath: string }>;
