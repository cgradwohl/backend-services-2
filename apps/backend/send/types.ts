import { Kinesis } from "aws-sdk";
import { DocumentClient } from "aws-sdk/clients/dynamodb";
import {
  Content,
  MessageChannels,
  MessageData,
  MessageMetadata,
  Recipient,
  RequestV2,
  TimeoutDateEpochSeconds,
  UTMMap,
} from "~/api/send/types";
import { IRenderedTemplatesMap } from "~/handlebars/template/render-templates";
import {
  IBrandColors,
  IBrandSettingsEmail,
  IBrandSnippets,
} from "~/lib/brands/types";
import { PricingPlan } from "~/lib/plan-pricing";
import {
  RouteNode,
  RouteNodeAddress,
  RouteTimeoutTable,
  RoutingStrategy,
  SendTimes,
} from "~/lib/send-routing";
import { TokensByProvider } from "~/lib/token-storage";
import { DeliveryHandlerParams } from "~/providers/types";
import {
  IBrand,
  IConfiguration,
  INotificationWire,
  IProfile,
  ITenant,
  NotificationCategory,
} from "~/types.api";
import {
  CourierRenderOverrides,
  TenantRouting,
  TenantScope,
} from "~/types.internal";
import { IProfilePreferences } from "~/types.public";

export type ApiVersion = "2019-04-01" | "2021-11-01";

export interface IRequest {
  apiVersion: ApiVersion;
  dryRunKey: TenantRouting;
  idempotencyKey: string | undefined;
  jobId?: string;
  params?: Record<"originalRequestId", string>;
  request: RequestV2;
  requestId: string;
  scope: TenantScope;
  source?: string;
  translated?: boolean;
}

export interface IMessage {
  apiVersion: ApiVersion;
  idempotencyKey: string | undefined;
  jobId?: string;
  messageId: string;
  message: RequestV2["message"];
  requestId: string;
  sequenceId?: string;
  sequenceActionId?: string;
  nextSequenceActionId?: string;
  source?: string;
}

export interface ISendProviderPayload {
  command: "send";
  channel: string;
  channelId: string;
  contextFilePath: string;
  dryRunKey: string;
  messageId: string;
  outputFilePath: string;
  messageFilePath: string;
  configurationId: string;
  tenantId: string;
  requestId: string;
  address?: RouteNodeAddress;
  times?: SendTimes;
  translated?: boolean;
}

export interface IRenderProviderPayload {
  command: "render";
  channel: string;
  channelId?: string;
  dryRunKey: string;
  contextFilePath: string;
  messageId: string;
  messageFilePath: string;
  configurationId: string;
  requestId: string;
  tenantId: string;
  address?: RouteNodeAddress;
  times?: SendTimes;
  shouldVerifyRequestTranslation: boolean;
  translated: boolean;
}

export type Command =
  | "request"
  | "accept"
  | "ad-hoc-list"
  | "send-audiences-member"
  | "send-audiences"
  | "list-pattern"
  | "list"
  | "prepare"
  | "render"
  | "route"
  | "send"
  | "sequence";

export enum SendActionCommands {
  Request = "request",
  Accept = "accept",
  Prepare = "prepare",
  Route = "route",
  Render = "render",
  Send = "send",
}

export interface ActionCommands {
  "ad-hoc-list": (action: IAdHocListAction) => Promise<void>;
  "send-audiences-member": (
    action: ISendAudiencesMemberAction
  ) => Promise<void>;
  "send-audiences": (action: ISendAudiencesAction) => Promise<void>;
  "list-pattern": (action: IListPatternAction) => Promise<void>;
  accept: (action: IAcceptAction) => Promise<void>;
  list: (action: IListAction) => Promise<void>;
  prepare: (action: IPrepareAction) => Promise<void>;
  request: (action: IRequestAction) => Promise<void>;
  route: (action: IRouteAction) => Promise<void>;
}

export interface IRetryableAction {
  retryCount?: number;
}

export interface IAction extends IRetryableAction {
  command: Command;
  dryRunKey: string;
  requestId: string;
  tenantId: string;
  shouldVerifyRequestTranslation?: boolean;
  translated?: boolean;
  shouldUseInboundSegmentEventKinesis?: boolean;
}

export interface IRequestAction extends IAction {
  apiVersion: ApiVersion;
  dryRunKey: TenantRouting;
  command: "request";
  requestFilePath: string;
  scope: TenantScope;
  source: string;
  shouldVerifyRequestTranslation?: boolean;
  translated?: boolean;
}

export interface ISequenceAction extends IAction {
  command: "sequence";
  sequenceId: string;
  sequenceActionId: string;
  tenantId: string;
}

export interface IAcceptAction extends IAction {
  command: "accept";
  shouldVerifyRequestTranslation?: boolean;
  translated?: boolean;
}

export interface IListAction extends IAction {
  command: "list";
  exclusiveStartKey?: DocumentClient.Key;
}

export interface ISendAudiencesAction extends IAction {
  audienceId: string;
  command: "send-audiences";
  cursor?: string;
}

export interface ISendAudiencesMemberAction extends IAction {
  command: "send-audiences-member";
  memberId: string;
}

export interface IListPatternAction extends IAction {
  command: "list-pattern";
  exclusiveStartKey?: DocumentClient.Key;
}

export interface IAdHocListAction extends IAction {
  command: "ad-hoc-list";
}

export interface IPrepareAction extends IAction {
  command: "prepare";
  messageId: string;
  messageFilePath: string;
  scheduleJobId?: string;
  shouldVerifyRequestTranslation?: boolean;
  translated?: boolean;
}

export interface IRouteAction extends IAction {
  command: "route";
  contextFilePath: string;
  messageId: string;
  messageFilePath?: string;
  times?: SendTimes;
  failedAddress?: RouteNodeAddress;
  shouldVerifyRequestTranslation?: boolean;
  translated?: boolean;
}

export interface IActionRecord extends Kinesis.Types.PutRecordInput {
  Data: IAction;
  PartitionKey: string; // generatePartition(requestId | messageId | actionId) ->
  StreamName: string;
}

export interface IRenderRecord extends Kinesis.Types.PutRecordInput {
  Data: IRenderProviderPayload;
  PartitionKey: string; // generatePartition(requestId | messageId | actionId) ->
  StreamName: string;
}

export interface ISendRecord extends Kinesis.Types.PutRecordInput {
  Data: ISendProviderPayload;
  PartitionKey: string; // generatePartition(requestId | messageId | actionId) ->
  StreamName: string;
}

export interface ISequenceRecord extends Kinesis.Types.PutRecordInput {
  Data: ISequenceAction;
  PartitionKey: string;
  StreamName: string;
}

export type ActionService = (tenantId: string) => {
  emit: <T extends IAction>(action: T) => Promise<void>;
  emitActions: <T extends IAction>(actions: T[]) => Promise<void>;
};

export type RenderService = (tenantId: string) => {
  emit: (payload: IRenderProviderPayload) => Promise<void>;
};

export type RequestService = (tenantId: string) => {
  create: (payload: IRequest) => Promise<{ filePath: string }>;
};
export type MessageService = (tenantId: string) => {
  create: (payload: {
    message: IMessage;
    shouldVerifyRequestTranslation?: boolean;
    translated?: boolean;
  }) => Promise<{ filePath: string; message: IMessage }>;
  get: (payload: { filePath: string }) => Promise<IMessage | undefined | null>;
};

export type SendService = (tenantId: string) => {
  emit: (payload: ISendProviderPayload) => Promise<void>;
};

export type ContextService = (tenantId: string) => {
  create: (payload: {
    context: ISendMessageContext;
    messageId: string;
  }) => Promise<{ filePath: string }>;
  get: (payload: { filePath: string }) => Promise<ISendMessageContext>;
};

export type OutputService = (tenantId: string) => {
  create: (requestId: string, context: ISendMessageContext) => Promise<void>;
  get: (requestId: string, messageId: string) => Promise<ISendMessageContext>;
};

// tslint:disable-next-line: no-empty-interface
export interface IProviderConfiguration extends IConfiguration {}

export interface IBrandContext extends IBrand {
  partials?: {
    [snippetName: string]: string;
  };
  locales?: {
    [key: string]: IBrand;
  };
}

export interface IMessageBrands {
  /** Can be default brand, message.brand_id, or undefined if the notification specified brand disabled */
  main?: IBrandContext;
  /** brands of message.channels[channel].brand_id */
  channels?: {
    [channelName: string]: IBrandContext;
  };
}

export type AdHocListMessageRecipient = Recipient & { ad_hoc_list_id: string };
export interface ISendMessageContext {
  category?: NotificationCategory;
  content: Content | INotificationWire;
  channels?: MessageChannels;
  overrides?: {
    channels?: Record<string, any>;
    providers?: Record<string, any>;
    brand?: {
      snippets?: IBrandSnippets;
      settings?: {
        colors?: IBrandColors;
        email?: IBrandSettingsEmail;
      };
    };
  };
  data?: MessageData; // Should this be here or in variableData? Probably not both as it is now.
  dryRunKey?: TenantRouting;
  environment: "production" | "test";
  metadata?: MessageMetadata;
  preferences?: IProfilePreferences;
  profile: IProfile; // Should this be here or in variableData? Probably not both as it is now.
  providers: IProviderConfiguration[];
  routingTree?: RouteNode;
  timeouts?: RouteTimeoutTable;
  scope: string;
  tenant: ITenant;
  /** Note: Only available to v2 pipeline. We should add snake_case aliases for these values */
  variableData: {
    data?: MessageData;
    courier: CourierRenderOverrides;
    messageId: string;
    event?: string; // In v1 this is templateId, V2 message.metadata.event ?? message.template
    template?: string; // V2 send specified template
    profile: IProfile;
    recipient?: string; // Alias for user_id for v1 back compat
    openTrackingId?: string;
    unsubscribeTrackingId?: string;
    urls?: {
      opened: string | null;
      unsubscribe: string | null;
      preferences?: string | null;
    };
    utmMap?: UTMMap;
  };
  brands: IMessageBrands;
}

export interface IRenderedOutput {
  deliveryHandlerParams: DeliveryHandlerParams;
  renderedTemplates: IRenderedTemplatesMap;
  pricingPlan?: PricingPlan;

  // Duplicated data from ISendMessage context to reduce latency. TODO: Remove tree
  tree?: RouteNode;
  timeouts?: RouteTimeoutTable;
}

export type PublishedState = "published" | "draft" | "submitted";

export type Environment = "production" | "test";
