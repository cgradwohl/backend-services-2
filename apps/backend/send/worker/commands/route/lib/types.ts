import { IProviderConfiguration } from "~/send/types";

export type MatchingProvidersByChannel = {
  [channel: string]: IProviderConfiguration[];
};
