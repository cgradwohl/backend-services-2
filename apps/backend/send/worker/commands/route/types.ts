import { MessageStatusReason } from "~/lib/message-service/types";
import {
  IChannelConfiguration,
  IChannelProvider,
  IChannelProviderConfiguration,
  IConditionalConfig,
} from "~/types.api";

export interface IChannelSummary {
  channel: string;
  reason: string;
  selected: boolean;
}

export interface IUndeliverableEventLog {
  type: MessageStatusReason;
  reason: string;
  data: Record<string, unknown>;
}

export interface IRoutingSummary {
  channel: string;
  provider?: string;
  selected: boolean;
  timedout?: boolean;
  blockIds?: string[] | null;
  config?: IChannelConfiguration | IChannelProviderConfiguration | undefined;
  id?: string;
  reason?: string;
  providers?: IChannelProvider[] | null;
  taxonomy?: string;
  disabled?: boolean;
  label?: string;
  conditional?: IConditionalConfig;
  key?: string;
  // Provider Configuration ID
  configurationId?: string;
}
