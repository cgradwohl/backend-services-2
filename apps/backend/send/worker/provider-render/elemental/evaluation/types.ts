import { ElementalNodeIR } from "~/api/send/types";

export type ElementalRefs = { [refId: string]: ElementalNodeIR };
