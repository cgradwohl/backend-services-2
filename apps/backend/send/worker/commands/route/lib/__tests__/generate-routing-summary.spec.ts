import { RoutingStrategy, RoutingSummary } from "~/lib/send-routing/types";
import {
  applyNotificationToRoutingSummary,
  generateRoutingSummary,
} from "../generate-routing-summary";

const providerConfigs: any = [
  {
    json: {
      domain: "sandboxb1d4dd0eb044497987e3c14f3c7c4db6.mailgun.org",
      apiKey: "4cea38609bdedcd992c18fd595a55a6d-7b8c9ba8-91e9abcd",
      fromAddress: "cmoney@example.com",
      provider: "apn",
    },
    id: "f2661ebd-dd91-417c-8b37-a9870fd0cbf4",
  },
  {
    json: {
      apiKey: "SG.gf.df",
      fromAddress: "chris@courier.com",
      checkDeliveryStatus: false,
      provider: "airship",
    },
    id: "a0e317ae-b30b-482f-a503-5e772e234f65",
  },
  {
    json: {
      accountSid: "wetwetgs",
      authToken: "tewthfdadfdf",
      messagingServiceSid: "greswetwtre",
      provider: "twilio",
    },
    id: "7593c6b6-f241-47f0-85b4-b5386cd60086",
  },
  {
    json: {
      accountSid: "wetwetgs",
      authToken: "tewthfdadfdf",
      messagingServiceSid: "greswetwtre",
      provider: "vonage",
    },
    id: "7593c6b6-f241-47f0-85b4-vonage",
  },
  {
    json: {
      accountSid: "wretewrtwq",
      authToken: "wtewetsgfdgssdgf",
      messagingServiceSid: "sdfgsdfgfdsdssdfg",
      provider: "mailjet",
    },
    id: "7593c6b6-f241-47f0-85b4-b5386cd600er6",
  },
  {
    json: {
      accountSid: "GSGDSewrewrewr",
      authToken: "gdfsgferger345sfa",
      messagingServiceSid: "sdfsggfdsdfsgfsdfger",
      provider: "sendgrid",
    },
    id: "2321",
  },
  {
    json: {
      accountSid: "gsdfds4353454",
      authToken: "324323423423",
      messagingServiceSid: "adsfasadfaswewr",
      provider: "slack",
    },
    id: "slack2e1231",
  },
];

describe("generate route summary", () => {
  describe("generateRoutingSummary", () => {
    it("should return a valid summary from a basic strategy", async () => {
      const strategy: RoutingStrategy = {
        routing: {
          method: "all",
          channels: ["sms", "email"],
        },
        channels: {},
        providers: {},
      };

      const summary = await generateRoutingSummary({
        strategy,
        providerConfigs,
        params: {
          data: {},
          profile: { phone_number: "1234567890", email: "drew@ycourier.com" },
        },
      });

      expect(summary).toMatchInlineSnapshot(`
        Array [
          Object {
            "channel": "sms",
            "configurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
            "provider": "twilio",
            "selected": true,
            "taxonomy": "direct_message:sms:twilio",
          },
          Object {
            "channel": "email",
            "configurationId": "7593c6b6-f241-47f0-85b4-b5386cd600er6",
            "provider": "mailjet",
            "selected": true,
            "taxonomy": "email:mailjet",
          },
        ]
      `);
    });

    it("should handle a single method strategy", async () => {
      const strategy: RoutingStrategy = {
        routing: {
          method: "single",
          channels: ["sms", "email"],
        },
        channels: {},
        providers: {},
      };

      const summary = await generateRoutingSummary({
        strategy,
        providerConfigs,
        params: {
          data: {},
          profile: { phone_number: "1234567890", email: "drew@ycourier.com" },
        },
      });

      expect(summary).toMatchInlineSnapshot(`
        Array [
          Object {
            "channel": "sms",
            "configurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
            "provider": "twilio",
            "selected": true,
            "taxonomy": "direct_message:sms:twilio",
          },
        ]
      `);
    });

    it("should respect channel provider restrictions", async () => {
      const strategy: RoutingStrategy = {
        routing: {
          method: "all",
          channels: ["sms", "email"],
        },
        channels: {
          sms: {
            providers: ["vonage"],
          },
          email: {
            providers: ["sendgrid"],
          },
        },
        providers: {},
      };

      const summary = await generateRoutingSummary({
        strategy,
        providerConfigs,
        params: {
          data: {},
          profile: { phone_number: "1234567890", email: "drew@ycourier.com" },
        },
      });

      expect(summary).toMatchInlineSnapshot(`
        Array [
          Object {
            "channel": "sms",
            "configurationId": "7593c6b6-f241-47f0-85b4-vonage",
            "provider": "vonage",
            "selected": true,
            "taxonomy": "direct_message:sms:vonage",
          },
          Object {
            "channel": "email",
            "configurationId": "2321",
            "provider": "sendgrid",
            "selected": true,
            "taxonomy": "email:sendgrid",
          },
        ]
      `);
    });

    it("should respect channel conditionals", async () => {
      const strategy: RoutingStrategy = {
        routing: {
          method: "all",
          channels: ["sms", "email"],
        },
        channels: {
          sms: {
            if: "data.locale === 'en-US'",
          },
          email: {
            if: "data.locale === 'eu-FR'",
          },
        },
        providers: {},
      };

      const summary = await generateRoutingSummary({
        strategy,
        providerConfigs,
        params: {
          data: { locale: "en-US" },
          profile: { phone_number: "1234567890", email: "drew@ycourier.com" },
        },
      });

      expect(summary).toMatchInlineSnapshot(`
        Array [
          Object {
            "channel": "sms",
            "configurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
            "provider": "twilio",
            "selected": true,
            "taxonomy": "direct_message:sms:twilio",
          },
          Object {
            "channel": "email",
            "provider": "",
            "reason": "Channel conditional failed",
            "selected": false,
            "type": "FILTERED",
          },
        ]
      `);
    });

    it("should respect provider conditionals", async () => {
      const strategy: RoutingStrategy = {
        routing: {
          method: "all",
          channels: ["sms", "email"],
        },
        channels: {
          sms: {
            providers: ["twilio"],
          },
          email: {
            providers: ["mailjet"],
          },
        },
        providers: {
          twilio: {
            if: "data.locale === 'en-US'",
          },
          mailjet: {
            if: "data.locale === 'eu-FR'",
          },
        },
      };

      const summary = await generateRoutingSummary({
        strategy,
        providerConfigs,
        params: {
          data: { locale: "en-US" },
          profile: { phone_number: "1234567890", email: "drew@courier.com" },
        },
      });

      expect(summary).toMatchInlineSnapshot(`
        Array [
          Object {
            "channel": "sms",
            "configurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
            "provider": "twilio",
            "selected": true,
            "taxonomy": "direct_message:sms:twilio",
          },
          Object {
            "channel": "email",
            "provider": "",
            "reason": "No remaining configured providers for channel or channel is invalid",
            "selected": false,
            "type": "NO_PROVIDERS",
          },
        ]
      `);
    });

    it("should allow providers to be channels", async () => {
      const strategy: RoutingStrategy = {
        routing: {
          method: "all",
          channels: ["mailjet"],
        },
        channels: {},
        providers: {},
      };

      const summary = await generateRoutingSummary({
        strategy,
        providerConfigs,
        params: {
          data: {},
          profile: { email: "drew@courier.com" },
        },
      });

      expect(summary).toMatchInlineSnapshot(`
        Array [
          Object {
            "channel": "mailjet",
            "configurationId": "7593c6b6-f241-47f0-85b4-b5386cd600er6",
            "provider": "mailjet",
            "selected": true,
            "taxonomy": "email:mailjet",
          },
        ]
      `);
    });

    it("should handle misspelled channels and providers without puking", async () => {
      const strategy: RoutingStrategy = {
        routing: {
          method: "all",
          channels: ["mailet", "eml"],
        },
        channels: {},
        providers: {},
      };

      const summary = await generateRoutingSummary({
        strategy,
        providerConfigs,
        params: {
          data: {},
          profile: { email: "drew@courier.com" },
        },
      });

      expect(summary).toMatchInlineSnapshot(`
        Array [
          Object {
            "channel": "mailet",
            "provider": "",
            "reason": "No remaining configured providers for channel or channel is invalid",
            "selected": false,
            "type": "NO_PROVIDERS",
          },
          Object {
            "channel": "eml",
            "provider": "",
            "reason": "No remaining configured providers for channel or channel is invalid",
            "selected": false,
            "type": "NO_PROVIDERS",
          },
        ]
      `);
    });

    it("should call channel handles", async () => {
      const strategy: RoutingStrategy = {
        routing: {
          method: "all",
          channels: ["sms"],
        },
        channels: {},
        providers: {},
      };

      const summary = await generateRoutingSummary({
        strategy,
        providerConfigs,
        params: {
          data: {},
          profile: {},
        },
      });

      expect(summary).toMatchInlineSnapshot(`
        Array [
          Object {
            "channel": "sms",
            "provider": "twilio",
            "reason": "Message is missing the minimum data required to send with this provider. Please check the message's profile and data properties and try again.",
            "selected": false,
            "type": "MISSING_PROVIDER_SUPPORT",
          },
          Object {
            "channel": "sms",
            "provider": "vonage",
            "reason": "Message is missing the minimum data required to send with this provider. Please check the message's profile and data properties and try again.",
            "selected": false,
            "type": "MISSING_PROVIDER_SUPPORT",
          },
        ]
      `);
    });

    it("should pass tokens to handles and produce correct summaries", async () => {
      const strategy: RoutingStrategy = {
        routing: {
          method: "all",
          channels: ["apn"],
        },
        channels: {},
        providers: {},
      };

      const apnSummaries = await Promise.all([
        generateRoutingSummary({
          strategy,
          providerConfigs,
          params: {
            data: {},
            profile: {},
          },
        }),
        generateRoutingSummary({
          strategy,
          providerConfigs,
          tokens: { apn: [{ token: "hello!" } as any] },
          params: {
            data: {},
            profile: {},
          },
        }),
      ]);

      expect(apnSummaries).toMatchInlineSnapshot(`
        Array [
          Array [
            Object {
              "channel": "apn",
              "provider": "apn",
              "reason": "Message is missing the minimum data required to send with this provider. Please check the message's profile and data properties and try again.",
              "selected": false,
              "type": "MISSING_PROVIDER_SUPPORT",
            },
          ],
          Array [
            Object {
              "channel": "apn",
              "configurationId": "f2661ebd-dd91-417c-8b37-a9870fd0cbf4",
              "provider": "apn",
              "selected": true,
              "taxonomy": "push:apn",
            },
          ],
        ]
      `);
    });

    it("should respect channel routing_method", async () => {
      const strategy: RoutingStrategy = {
        routing: {
          method: "single",
          channels: ["push"],
        },
        channels: {
          push: {
            routing_method: "all",
            providers: ["apn", "airship"],
          },
        },
        providers: {},
      };

      const summary = await generateRoutingSummary({
        strategy,
        providerConfigs,
        params: {
          data: {},
          profile: {
            apn: {
              token: "YOUR TOKEN",
            },
            airship: {
              audience: {
                named_user: "MEEEE",
              },
              device_types: ["ios"],
            },
          },
        },
      });

      expect(summary).toMatchInlineSnapshot(`
        Array [
          Object {
            "channel": "push",
            "configurationId": "f2661ebd-dd91-417c-8b37-a9870fd0cbf4",
            "provider": "apn",
            "selected": true,
            "taxonomy": "push:apn",
          },
          Object {
            "channel": "push",
            "configurationId": "a0e317ae-b30b-482f-a503-5e772e234f65",
            "provider": "airship",
            "selected": true,
            "taxonomy": "push:airship",
          },
        ]
      `);
    });

    it("should not duplicate sms providers", async () => {
      const strategy: RoutingStrategy = {
        routing: {
          method: "all",
          channels: ["sms"],
        },
        channels: {
          sms: {
            providers: ["twilio"],
          },
        },
        providers: {},
      };

      const summary = await generateRoutingSummary({
        strategy,
        providerConfigs,
        params: {
          data: {},
          profile: { phone_number: "1234567890" },
        },
      });

      expect(summary).toMatchInlineSnapshot(`
        Array [
          Object {
            "channel": "sms",
            "configurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
            "provider": "twilio",
            "selected": true,
            "taxonomy": "direct_message:sms:twilio",
          },
        ]
      `);
    });

    it("should handle sub-routing strategies", async () => {
      const strategy: RoutingStrategy = {
        routing: {
          method: "all",
          channels: [
            {
              method: "single",
              channels: ["sms", "push"],
            },
            {
              method: "single",
              channels: ["email", "slack"],
            },
          ],
        },
        channels: {},
        providers: {},
      };

      const summary = await generateRoutingSummary({
        strategy,
        providerConfigs,
        params: {
          data: {},
          profile: {
            email: "drew@courier.com",
            phone_number: "1234567890",
            apn: {
              token: "YOUR TOKEN",
            },
          },
        },
      });

      expect(summary).toMatchInlineSnapshot(`
        Array [
          Object {
            "channel": "sms",
            "configurationId": "7593c6b6-f241-47f0-85b4-b5386cd60086",
            "provider": "twilio",
            "selected": true,
            "taxonomy": "direct_message:sms:twilio",
          },
          Object {
            "channel": "email",
            "configurationId": "7593c6b6-f241-47f0-85b4-b5386cd600er6",
            "provider": "mailjet",
            "selected": true,
            "taxonomy": "email:mailjet",
          },
        ]
      `);
    });
  });

  describe("applyNotificationToRoutingSummary", () => {
    const summary: RoutingSummary[] = [
      {
        channel: "email",
        configurationId: "b46abb6b-e99e-473b-927a-7d9843ec0442",
        provider: "mailjet",
        selected: true,
        taxonomy: "email:mailjet",
      },
    ];
    const notification: any = {
      json: {
        channels: {
          always: [
            {
              channel: "email",
              channelLabel: "Email",
              provider: "mailjet",
              selected: true,
              key: "mailjet",
              configurationId: "b46abb6b-e99e-473b-927a-7d9843ec0442",
              config: {
                email: {
                  emailSubject: "My Test Subject {name}",
                  emailTemplateConfig: {
                    templateName: "none",
                    topBarColor: "#9121C2",
                  },
                  renderPlainText: false,
                },
                locales: {},
              },
              conditional: {
                filters: [
                  {
                    source: "data",
                    property: "do_email",
                    value: "true",
                    operator: "EQUALS",
                    id: "af4fc93d-4321-4737-9556-e8954d33a9ad",
                  },
                ],
                logicalOperator: "and",
                behavior: "hide",
              },
              blockIds: ["92e08808-8883-4034-be2f-d372df5db3c1"],
              id: "6721af5a-7d56-4906-a021-6f30fe84253a",
              providers: [
                {
                  key: "mailjet",
                  configurationId: "b46abb6b-e99e-473b-927a-7d9843ec0442",
                  config: {},
                  conditional: {
                    filters: [
                      {
                        source: "data",
                        property: "dont",
                        value: "true",
                        operator: "EQUALS",
                        id: "278a9468-8159-4518-8554-8212d7e3c9fd",
                      },
                    ],
                    logicalOperator: "and",
                    behavior: "hide",
                  },
                },
              ],
              taxonomy: "email:*",
              disabled: false,
              label: "Email",
            },
          ],
        },
      },
    };

    it("should leave routing summary alone if no notification", () => {
      const result = applyNotificationToRoutingSummary(summary);
      expect(result).toEqual(summary);
    });

    it("should add relevant information if notification", () => {
      const result = applyNotificationToRoutingSummary(summary, notification);
      expect(result).toMatchSnapshot();
    });
  });
});
