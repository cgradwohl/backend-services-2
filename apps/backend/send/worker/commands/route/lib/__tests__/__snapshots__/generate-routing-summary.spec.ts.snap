// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate route summary applyNotificationToRoutingSummary should add relevant information if notification 1`] = `
Array [
  Object {
    "channel": "email",
    "configurationId": "b46abb6b-e99e-473b-927a-7d9843ec0442",
    "id": "6721af5a-7d56-4906-a021-6f30fe84253a",
    "provider": "mailjet",
    "selected": true,
    "taxonomy": "email:mailjet",
  },
]
`;
